### DISCUSSION TITLE
Cache global or per panel?

### DISCUSSION THREAD

**Discussion by tillsteinbach** [grafana/grafana-json-datasource#261] [Q&A]
I was expecting that the datasource has a central cache and not a cache per panel. My tries let me assume that is not the case.
I have a dashboard that uses the same URL in approx. 10 dashboards resulting in fetching the same data 10 times.

Is that a bug? Is that expected behaviour? If it is expected, can I ask for a new feature to share the cache between different panels (or even dashboards)?

**Comment by gianluca-nitti** [comment]
+1 for the dashboard-wide cache, it would be useful to reduce the load on the server in situations where one needs to feed data from the same API call into different panels, each with its own JSONPath field specifications. Unfortunately however I'm not sure if it's feasible since if I understand correctly Grafana loads data for the various panels in a dashboard in parallel. Thus it's not like it loads panel A first, making the API call, then loads panel B which can use the cached result, but loads both A and B in parallel thus making the request twice even if it's identical. Maybe someone with deeper knowledge of Grafana internals can confirm this?

**Comment by marcusolsson** [comment]
I believe this is how most data sources in Grafana work. Have you tried the [Dashboard data source](https://grafana.com/blog/2020/10/14/learn-grafana-share-query-results-between-panels-to-reduce-load-time/)?

I'm afraid that to support something like this, we'd need to move the plugin server-side, which is a non-trivial feature, and (in this case) requires a major rewrite.

In the meantime, could you potentially set up NGINX to cache things across queries?

**Reply by tillsteinbach** [reply]
The problem with the Dashboard data source is: I cannot share the exact query as it is slightly different for each panel. I just want to share the http request made from the json-datasource, because it is the same for every panel. I can setup a server side cache but still it will have plenty of unnecessary fetches of the json-data.

**Reply by marcusolsson** [reply]
> it will have plenty of unnecessary fetches of the json-data

From Grafana to your proxy/cache, yes, but at least it'll take some of the load of the API.

I agree it's not ideal. If someone wants to look into this, I'd be happy to accept a pull request.

**Reply by gianluca-nitti** [reply]
Thank you @marcusolsson for the explanation. I also tried the dashboard data source like @tillsteinbach and have the same problem: it allows to reuse the same **processed** results, i.e. after applying the JSONPath expressions, but not the raw API response.

So, just to clarify, if an API endpoint returns a large JSON which I want to map in different ways to different panels, such as `{"dataForPanel1": ..., "dataForPanel2": ...}` in my understanding it's not possible to reuse the same response. An interesting way to work around this could be using Grafana's [transforms](https://grafana.com/docs/grafana/next/panels/transform-data/), however they are not powerful enough for my use case as they don't support something like JSONPath or custom JS.

**Reply by marcusolsson** [reply]
Thanks for sharing your experiences. I originally put in the cache to avoid hitting rate limits when you're editing queries, for example since the auto-completion needs to query the JSON object as you type. I'm open to improving this, though I can't guarantee when I'll have the time to look into it myself.
